### DISCUSSION TITLE
Tx build fails when specifying utxo to be consumed (object has no attribute 'output')

### DISCUSSION THREAD

**Discussion by ajuggler** [Python-Cardano/pycardano#245] [Q&A] [answered]
In this example I try to build a transaction where the UTxO to be consumed has been explicitly specified.  It doesn't work, and throws the error
```shell
    selected_amount += i.output.amount
                       ^^^^^^^^
    AttributeError: 'TransactionInput' object has no attribute 'output'
```
which I interpret as PyCardano being unable to query the blockchain for the output amount associated with the provided UTxO.  Have I found a bug in PyCardano?  Here is the code:

```python
import os
from pycardano import (
    Transaction,
    TransactionBody,
    TransactionInput,
    TransactionOutput,
    TransactionBuilder,
    BlockFrostChainContext,
    Value,
    Network,
    Address,
)
from blockfrost import ApiUrls

context = BlockFrostChainContext("my-consealed-preview-project-id", base_url=ApiUrls.preview.value)
NETWORK = Network.TESTNET

# ADDRESSES

wallet2_address = 'addr_test1vznjse8fycd7dzw9usre7vka0rwjcqszuu7a26pkdrw0g8sczt0l5'

# INPUTS

tx_in = TransactionInput.from_primitive(
    ['4a16c4f23a1855392496650deac62eb6c1fa1887f7a2e59ee2a3a3c7b67fa733', 1])

# OUTPUTS

tx_out = TransactionOutput(address=wallet2_address, amount=Value(5000000))

# BUILD TRANSACTION                                                          

tx_builder = TransactionBuilder(context)

tx_builder.add_input(tx_in);

tx_builder.add_output(tx_out);

tx_builder.build()
```

**Comment by nielstron** [comment]
Please try `i.input.amount` instead. The error displayed is an attribute error and is related to using the wrong object to achieve what you want to do, not a bug in PyCardano.

**Comment by cffls** [comment, marked as answer]
`tx_in` defined in your script should be a [`UTxO`](https://pycardano.readthedocs.io/en/latest/api/pycardano.transaction.html#pycardano.transaction.UTxO) type instead of a `TransactionInput`.

**Comment by ajuggler** [comment]
As cffls' answer states, `tx_in` should be a UTxO, which is an object with `input` and `output` attributes.  With this in mind, I figured out the correct code:

```python
# Recall that my objective is to consume a specific UTxO, now determined by variable 'tx_in_input'

# ADDRESSES

wallet1_address = 'addr_test1vq8y2nlht8ek0elatvksa5ctg6u4eatrj6hhqz3ln9ca0gsn3cnam'

wallet2_address = 'addr_test1vznjse8fycd7dzw9usre7vka0rwjcqszuu7a26pkdrw0g8sczt0l5'

# INPUTS

tx_in_input = TransactionInput.from_primitive(
    ['4a16c4f23a1855392496650deac62eb6c1fa1887f7a2e59ee2a3a3c7b67fa733', 1])

tx_in = next(
    (
        u for u in context.utxos(wallet1_address)
        if u.input == tx_in_input
    ),
    None
)

# OUTPUTS

tx_out = TransactionOutput(address=wallet2_address, amount=Value(5000000))

# BUILD TRANSACTION                                                          

tx_builder = TransactionBuilder(context)

tx_builder.add_input(tx_in);

tx_builder.add_output(tx_out);

tx_builder.build()
```

The only drawback is that it is not enough to know `tx_in_input`, as also the wallet address where it is locked (*wallet1* in this case) is also needed.  I think there is no way around this.

**Reply by cffls** [reply]
You can simplify the script by providing `wallet1_address` to method `add_input_address` in txbuilder, something like this:

```python
tx_builder.add_input_address(wallet1_address)
tx_builder.add_output(tx_out);
tx_builder.build()
```

Then the tx builder will search for all utxos under this wallet address when building the transaction.

**Reply by ajuggler** [reply]
But I don't think this would guarantee that the transaction consumes the particular UTxO that I want it to consume.  (Why would I insist in picking a particular UTxO?  Maybe because this is part of a minting transaction and the minting policy enforces consumption of a particular UTxO in order to guarantee that an NFT is in fact unique.)

**Reply by cffls** [reply]
Okay, in that case you can list utxos using context and loop through all utxos until you see the one you want to spend:

```python
utxo_in = None
tx_in_input = TransactionInput.from_primitive(
    ['4a16c4f23a1855392496650deac62eb6c1fa1887f7a2e59ee2a3a3c7b67fa733', 1])
utxos = context.utxos(wallet1_address)
for utxo in utxos:
    if utxo.intput == tx_in_input
    utxo_in = utxo

tx_builder.add_input(utxo_in)
tx_builder.add_output(tx_out);
tx_builder.build()
```
